### DISCUSSION TITLE
"object (___) was not found in its object stream" error

### DISCUSSION THREAD

**Discussion by brandxntu** [pymupdf/PyMuPDF#1213] [Looking for help] [answered]
I am experiencing an issue with a handful of PDF files.

After opening these specific files with fitz.open(file) , I am greeted with a string of the following errors:

> mupdf: object (2065 0 R) was not found in its object stream
> mupdf: object (2068 0 R) was not found in its object stream
> mupdf: object (2071 0 R) was not found in its object stream
> mupdf: object (2075 0 R) was not found in its object stream

The program does not crash however.
Information can still be extracted from the file at this point including metadata, page.mediabox, and page.get_text().

However, when attempting to save the file using pdf.save() the program will crash.

> mupdf: object (1924 0 R) was not found in its object stream
> Traceback (most recent call last):
> 
>   File "<ipython-input-23-372ec84a2b66>", line 1, in <module>
>     pdf.save("file.pdf")
> 
> RuntimeError: object (1924 0 R) was not found in its object stream

Using metadata, I found that the problem files all had a metadata['creator'] value of "PDF-XChange Editor 7.0.326.1"

Any idea what is going on or how this issue could be fixed would be much appreciated.

**Comment by JorjMcKie** [comment, marked as answer]
These PDFs are severely damaged. Impossible to tell why from this end.
The message may sound cryptic, but that PDF viewer compresses PDF objects into so-called objects streams (object type "Objstm" to save space. These are objects that contain other object definitions. Here, MuPDF cannot find objects in some of these compressed data.

Try to clean the file via ``mutool clean -gggg damaged.pdf ...``. May not work if the problems are too severe. If it fails you may be able to at least rescue a few pages via `insert_pdf` using a new file.

**Reply by JorjMcKie** [reply]
@shmyyy - 

> fitz.open works on document - so means the document is not damaged.

No, wrong conclusion! It only means that some of the most severe errors either did not occur or were successfully repaired. Nothing else.

You PDF may still have all sorts of other errors: damaged  images, fonts, page tree, objects not found, ... you name it.
PDF is not a file format that can ensure consistency.

PyMuPDF does not "mess up" anything.

**Reply by JorjMcKie** [reply]
> So means it is not damaged ...

No, you cannot conclude this!
Displaying a PDF (paging through it with some PDF reader) does not necessarily access everything in the file. So you may not touch any messy part within the file.
Whereas saving the file accesses yet other parts - among them damaged pieces.

Please accept: **_your file is damaged!_**
You may be able to rescue (parts of) it in one or the other way, but there definitely are damages.

**Comment by renzo-montesa** [comment]
I have a similar case 
the error code is : object (2135 0 R) was not found in its object stream

**Reply by JorjMcKie** [reply]
Reiterating: Your PDF is damaged. If you have the latest PyMuPDF version you should be able to save it.

**Reply by renzo-montesa** [reply]
can i share you the code that im using PyMuPDF on. 
I am trying to batch process multiple pdf files. and obtaining annotations based on 
annotation_subject or subject in annotation_title or subject in annotation_name  . It checks all annotations and only keeps the annotation that the user inputted then saves it in a new pdf in a new directory.

Most of the files works properly and it saves the pdf with no error. but some of the pdfs that It process receives the error  object (2135 0 R) . It saves the file but when I open it its corrupted.

The original files before the process can be opened and checked in pdf viewer.  Only after its processed that it becomes corrupted. 

Im curious to know why its only corrupted when it is processed by the program and not corrupted when not processed. but this happens only to some files not all of it


import os
import fitz
import pandas as pd
import webbrowser

def process_pdfs(dir_path, comment_subjects_file):
    # Check if a directory is selected
    if not dir_path:
        print("No directory selected.")
        return

    # Check if the specified directory exists
    if not os.path.isdir(dir_path):
        print("Directory does not exist.")
        return

    # Check if the comment_subjects_file is provided
    if not comment_subjects_file:
        print("No comment subjects file provided."),
        return

    # Check if the comment_subjects_file is a valid CSV file
    if not comment_subjects_file.lower().endswith(".xlsx"):
        print("Invalid comment subjects file. Please provide a CSV file.")
        return

    # Load the comment subjects from the CSV file
    try:
        df = pd.read_excel(comment_subjects_file)
        print(df)
        df["Sub System Description"] = df["Sub System Description"].str.rstrip()
        df["Sub System Description"] = df["Sub System Description"].str.replace("/", " ").str.replace("\\", " ")
        comment_subjects = df["Sub System"].tolist()
        sub_system_description=df["Sub System Description"].tolist()
        # Remove trailing spaces in the "Sub System Description" column
        
    except Exception as e:
        print(f"Error loading comment subjects from the file: {e}")
        return
    # Check if comment_subjects is empty
    if not comment_subjects:
        print("No comment subjects found in the file.")
        return
    if not sub_system_description:
        print("No Sub System Description found in the file.")
        return

    print(dir_path)
  
    pdf_files = [file_name for file_name in os.listdir(dir_path) if file_name.lower().endswith(".pdf")]
    if len(pdf_files) == 0:
        print("No PDF files found in the directory ")
        return

    loop_counter=0
    for subject in comment_subjects:
        dir_path_subject=dir_path+"/"+subject
        print(subject)
        # Check if there are no PDF files in the directory
        # Create a folder for the modified PDFs
        description=sub_system_description[loop_counter]
        output_folder = os.path.join(dir_path, subject+"_"+description)
        os.makedirs(output_folder, exist_ok=True)
        loop_counter+=1

        # List to store PDFs that don't have any of the comment_subjects
        pdfs_without_comment = []
        file_code_error=['-----------SYNTAX ERRORS AHEAD----------']

        
        # Loop through all PDF files in the directory
        for file_name in pdf_files:
            pdfs_with_comment=[]
            file_name_error=[]
            file_path = os.path.join(dir_path, file_name)
            # Open the PDF file
            pdf_document = fitz.open(file_path)

            # Flag to indicate if the PDF has any comment with any of the comment_subjects
            has_comment = False

            for page in pdf_document:
                print(page)
                try:  
                    # Get all annotations (comments) on the page
                    annotations = page.annots()
                    comment_name=[]
                    # Loop through all comments and delete them
                    annot_counter=0
                    for annotation in annotations:
                    
                        annot_counter+=1
                        annotation_info = annotation.info
                        annotation_subject = str(annotation_info.get('subject', ''))
                        annotation_title=  str(annotation_info.get('title',''))
                        print(annotation_title)
                        annotation_name= str(annotation_info.get('name',''))
                       
                        has_matching_subject = False
        
                        if subject in annotation_subject or subject in annotation_title or subject in annotation_name :
                            has_matching_subject = True
                            if subject not in comment_name:
                                comment_name.append(subject)
                            
                        if has_matching_subject:
                            has_comment = True
                            
                        else:
                            page.delete_annot(annotation)
                except RuntimeError as e:
                    error_message = str(e)
                    if error_message.startswith("xref") and error_message.endswith("is not an annot of this page"):
                        xref_value = int(error_message.split()[1])
                        print(f"Error: xref {xref_value} is not an annot of this page")
                        error=f"Error: xref {xref_value} is not an annot of this page"
                        file_code_error.append(error)
                        file_name_error.append(file_name)
                        if file_name_error:
                            print("i am here")
                            notepad_files = os.path.join(output_folder, "files_skipped_because_error.txt")
                            with open(notepad_files, "w") as file:
                                for file_name in file_name_error:

                                    file.write(str(file_name) + "\n")
                                
                                for file_name in file_code_error:
                                    file.write(str(file_name) + "\n") 
                        continue  # Skip this xref and move to the next one
 
                    else:
                        # Handle other runtime errors if needed
                        print("An error occurred:", str(e))        
           

            # Create new filename with comment_subjects included
            new_file_name = file_name
            new_file_path = os.path.join(output_folder, new_file_name)

            # # Check if the PDF has any comment with any of the comment_subjects
            # if has_comment:
            #     # Save changes to new file
            #     pdf_document.save(new_file_path)
            # else:
            #     # Add the file name to the list of PDFs without comments
            #     pdfs_without_comment.append(file_name)


                        # Check if the PDF has any comment with any of the comment_subjects
            if has_comment:
                # Open the PDF file for manual inspection
                webbrowser.open(new_file_path)
            else:
                # Add the file name to the list of PDFs without comments
                pdfs_without_comment.append(file_name)
    

            # Close the PDF file
            pdf_document.close()

        # Print the list of PDF files without comments
        if pdfs_without_comment:
            notepad_file = os.path.join(output_folder, "files_without_comment.txt")
            with open(notepad_file, "w") as file:
                for file_name in pdfs_without_comment:

                    file.write(str(file_name) + "\n")
   
        print("Processing complete.")



# Get user input for directory path
directory_path = input("Enter the directory path: ")

# Get user input for file path
comment_subjects_file = input("Enter the file path for comment_subjects.csv: ")

process_pdfs(directory_path, comment_subjects_file)

**Reply by JorjMcKie** [reply]
This is a huge amount of code. I don't see how you save the changed documents: a lot depends on the save options - if you use too "soft" ones, a potential damage of the input may be propagated.
I believe you **are** using insufficient save options, because (Py-) MuPDF never creates those PDF objects which exclusively can cause that error message.

Message `RuntimeError: object (nnnn 0 R) was not found in its object stream` only occurs if there is reference to xref `nnnn` that is contained in a compressed "super"-object (object stream), but cannot be located in there. When using strong save options (garbage=3, deflate=True), object streams are dissolved and become normal PDF xrefs. Opening a file that has been saved like this never can generate this message.

What is your pymupdf version? You must use the most current one, because MuPDF has been too picky before that and raised an exception when it encountered that situation.

**Reply by JorjMcKie** [reply]
Another comment: A loop structured like this
```python
for annot in page.annots():
    ...
    page.delete_annot(annot)
```
is dangerous and will not work in older pymupdf versions, because you are altering the iterator `page.annots()` from within the loop that depends on it.
